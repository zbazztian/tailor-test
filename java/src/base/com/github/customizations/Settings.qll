import java

module Settings {
  bindingset[description]
  predicate parse(string description, string key, string value) {
    exists(int i | i = description.indexOf("=", 0, 0) |
      key = description.prefix(i).trim() and
      value = description.suffix(i + 1).trim()
    )
  }

  abstract class Provider extends int {
    bindingset[this]
    Provider() { any() }

    Provider getParent() { result = max(Provider p | p < this) }

    predicate rows(string key, string value) {
      if this.assign(key, _) then this.assign(key, value) else getParent().rows(key, value)
    }

    predicate assign(string key, string value) { none() }
  }

  string values(string key) { any(Provider p).rows(key, result) }

  string prioritizedValues(string key) { max(Provider p).rows(key, result) }
}
