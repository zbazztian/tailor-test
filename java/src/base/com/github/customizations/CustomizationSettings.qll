import com.github.customizations.Settings

module CustomizationSettings {
  class Defaults extends Settings::Provider {
    Defaults() { this = -2147483648 }

    override predicate assign(string key, string value) {
      key = "java.lenient_taintflow" and value = "false"
      or
      key = "java.local_sources" and value = "false"
    }
  }

  module java {
    predicate lenient_taintflow_enabled() {
      Settings::prioritizedValues("java.lenient_taintflow") = "true"
    }

    predicate local_sources_enabled() { Settings::prioritizedValues("java.local_sources") = "true" }

    int min_hash_iterations() {
      result = max(Settings::prioritizedValues("java.min_hash_iterations").toInt())
    }
  }
}
