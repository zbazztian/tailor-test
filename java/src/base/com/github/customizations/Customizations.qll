import com.github.customizations.CustomizationSettings
import semmle.code.java.dataflow.FlowSources
import semmle.code.java.dataflow.FlowSteps


class LocalIsRemote extends RemoteFlowSource {
  LocalIsRemote() {
    CustomizationSettings::java::local_sources_enabled() and
    this instanceof LocalUserInput
  }

  override string getSourceType() { result = "local source type" }
}
