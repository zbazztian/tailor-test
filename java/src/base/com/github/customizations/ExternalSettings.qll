import java
import com.github.customizations.Settings

external predicate external_customization_settings(string key, string value);

private module ExternalSettings {
  final class ExternalUserSettings extends Settings::Provider {
    ExternalUserSettings() { this = 2147483647 }

    override predicate assign(string key, string value) {
      external_customization_settings(key, value)
    }
  }
}
