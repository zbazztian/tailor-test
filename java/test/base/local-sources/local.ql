import com.github.customizations.Customizations
import com.github.customizations.Settings

class MyTailorSettings extends Settings::Provider {
  MyTailorSettings(){
    this = 0
  }

  override predicate assign(string key, string value) {
    key = "java.local_sources" and value = "true"
  }
}

from RemoteFlowSource rfs
select rfs
