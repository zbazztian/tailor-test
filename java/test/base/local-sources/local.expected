| Test.java:13:18:13:38 | getHostName(...) |
